### PR TITLE
Plugin specification

### PR DESCRIPTION
**Pull request by ash2k**

### CODE REVIEW

File: docs/design/plugins.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+      spec:
+        # need data from a-binding-secret but transformed
+        # e.g. only keys which start with "FOO_" when full names of the keys are not known in advance.
+```
+
+## Specification
+
+Smith plugins are [Go plugins](https://golang.org/pkg/plugin/). They are eagerly loaded at Smith startup to detect
+issues early. Names of plugins to load are passed via command line.
+Each plugin publishes a `Process(smith_v1.Resource, map[smith_v1.ResourceName]Dependency) (TransformResult, error)`
```
Comment (amckague) [inline docs/design/plugins.md]: s/TransformResult/ProcessResult

File: docs/design/plugins.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+// For reference:
+//
+//type Dependency struct {
+//	Spec    smith_v1.Resource
+//	Actual  runtime.Object
+//	Outputs []runtime.Object
+//}
+//
+//type ProcessResult struct {
+//	Object runtime.Object
```
Comment (amckague) [inline docs/design/plugins.md]: I take it runtime.Object is happy to be [object] and smith will handle the difference between [] and object?

File: docs/design/plugins.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+Smith plugins are [Go plugins](https://golang.org/pkg/plugin/). They are eagerly loaded at Smith startup to detect
+issues early. Names of plugins to load are passed via command line.
+Each plugin publishes a `Process(smith_v1.Resource, map[smith_v1.ResourceName]Dependency) (TransformResult, error)`
+function.
+
+When Smith comes across a resource with `type: plugin` and `pluginName: foobar` it invokes
+the plugin `foobar`. To do that it inspects all the dependencies of the resource specified in `dependsOn` attribute
+and fetches their outputs to include in the invocation along with the dependencies themselves. Smith needs to
+recognize resource group/version/kinds to be able to fetch the output (if there is one).
```
Comment (amckague) [inline docs/design/plugins.md]: Does this extend to allowing smith bundle objects to depend on objects from outside the bundle or is this the preferred mechanism to allow that to happen?

File: docs/design/plugins.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    pluginName: filter
+    pluginSpec:
+      apiVersion: servicecatalog.k8s.io/v1beta1
+      kind: ServiceInstance
+      name: b
+      spec:
+        filterByPrefix: "FOO_" # only keys which start with "FOO_"          
+```
+
+When the plugin `filter` is invoked, it returns the following object:
```
Comment (nilebox) [inline docs/design/plugins.md]: Would be nice for the sample plugin implementation above actually returning this result.

File: docs/design/plugins.md [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+//}
+//
+//type ProcessResult struct {
+//	Object runtime.Object
+//}
+
+func Process(resource smith_v1.Resource, dependencies map[smith_v1.ResourceName]smith_plugin.Dependency)  (smith_plugin.ProcessResult, error) {
+	// Do the processing
+	return smith_plugin.ProcessResult{
+		//Object: object literal here
```
Comment (nilebox) [inline docs/design/plugins.md]: Implement `filter` plugin there for a more intuitive example?

File: docs/design/plugins.md (hunk 2 above)
Comment (ash2k) [reply docs/design/plugins.md]: What is `[]` and `[object]`? I don't understand...

**Pull request merged**